File: sdk/python/packages/flet-core/src/flet_core/clipboard.py
import dataclasses
from typing import Any, Optional

from flet_core.control import Control
from flet_core.ref import Ref


@dataclasses.dataclass
class ClipboardData:
    ts: str
    d: Optional[str]


class Clipboard(Control):
    def __init__(
        self,
        ref: Optional[Ref] = None,
        data: Any = None,
    ):
        Control.__init__(
            self,
            ref=ref,
            data=data,
        )

    def _get_control_name(self):
        return "clipboard"

    def is_isolated(self):
        return True

    def set_data(self, data: str):
        self.invoke_method("set_data", {"data": data})

    def get_data(self) -> str:
        return self.invoke_method("get_data", wait_for_result=True)

    def get_data_async(self) -> str:
        return self.invoke_method_async("get_data", wait_for_result=True)

File: packages/flet/lib/src/services/clipboard.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:pasteboard/pasteboard.dart';

import '../flet_service.dart';
import '../utils/images.dart';

class ClipboardService extends FletService {
  ClipboardService({required super.control});

  @override
  void init() {
    super.init();
    debugPrint("ClipboardService(${control.id}).init: ${control.properties}");
    control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void update() {
    debugPrint("ClipboardService(${control.id}).update: ${control.properties}");
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("ClipboardService.$name($args)");
    switch (name) {
      case "set":
        await Clipboard.setData(ClipboardData(text: args["data"]));
      case "get":
        var data = await Clipboard.getData(Clipboard.kTextPlain);
        return data?.text;
      case "set_image":
        await Pasteboard.writeImage(convertToUint8List(args["data"]));
      case "get_image":
        return await Pasteboard.image;
      default:
        throw Exception("Unknown Clipboard method: $name");
    }
  }

  @override
  void dispose() {
    debugPrint("ClipboardService(${control.id}).dispose()");
    control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }
}
